### PR TITLE
Fix sync call to propose

### PR DESCRIPTION
**Pull request by tgrospic**
## Overview
This PR fixes sync version of propose function via API (continuation of #3352).

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 1]
```diff
@@ -105,14 +105,21 @@ object BlockAPI {
...
-                  case Some(Left(seqNum)) =>
-                    logSucess(s"Success: proposing block with seqNum ${seqNum}")
-                  case None => logDebug(s"Failure: another propose is in progress")
+            // Trigger propose
+            proposerResult <- triggerProposeF(casper, isAsync)
+            r <- proposerResult match {
+                  case ProposerEmpty =>
+                    logDebug(s"Propose failed: another propose is in progress")
+                  case ProposerFailure(status, seqNumber) =>
+                    logDebug(s"Propose failed: $status (seqNum $seqNumber)")
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: @zsluedem should these failed statuses be in some specific format for propose script? Really, it doesn't make sense to write the same message to logs and return to API.

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 2]
```diff
@@ -105,14 +105,21 @@ object BlockAPI {
...
+                  case ProposerStarted(seqNumber) =>
+                    logSucess(s"Propose started (seqNum $seqNumber)")
+                  case ProposerSuccess(_, block) =>
+                    // TODO: [WARNING] Format of this message is hardcoded in pyrchain when checking response result
+                    //  Fix to use structured result with transport errors/codes.
+                    // https://github.com/rchain/pyrchain/blob/a2959c75bf/rchain/client.py#L42
+                    val blockHashHex = block.blockHash.base16String
+                    logSucess(s"Success! Block $blockHashHex created and added.")
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:122]: Here we have excessive messages to logs. Proposer will also add to log when propose started and finished. I'll not address this as part of this PR.

File: casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala [hunk 3]
```diff
@@ -16,11 +16,25 @@ import coop.rchain.casper.util.rholang.RuntimeManager
...
-import coop.rchain.models.BlockHash.BlockHash
+
+sealed abstract class ProposerResult
+object ProposerEmpty                                                         extends ProposerResult
+final case class ProposerSuccess(status: ProposeStatus, block: BlockMessage) extends ProposerResult
+final case class ProposerFailure(status: ProposeStatus, seqNumber: Int)      extends ProposerResult
+final case class ProposerStarted(seqNumber: Int)                             extends ProposerResult
+
+object ProposerResult {
+  def empty: ProposerResult = ProposerEmpty
```
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala:31]: I'm curious why `empty`?  and not `busy`

**Review by nzpr — APPROVED** [review]
Just one question, but LGTM

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala (hunk 1 above)
Comment (zsluedem) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: > Really, it doesn't make sense to write the same message to logs and return to API.

Agree 100%. But that's how it is right now.

> should these failed statuses be in some specific format for propose script?

Yes.

File: casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala (hunk 3 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala:31]: @nzpr I was just following what kind of information has this constructor. For busy I'd expect to have info on what is processing like seq number, deploy, current progress, ...

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 4]
```diff
@@ -105,14 +105,21 @@ object BlockAPI {
...
-                  case Some(Right(blockHash)) =>
-                    logSucess(s"Success! Block ${blockHash.base16String} created and added.")
-                  case Some(Left(seqNum)) =>
-                    logSucess(s"Success: proposing block with seqNum ${seqNum}")
-                  case None => logDebug(s"Failure: another propose is in progress")
+            // Trigger propose
+            proposerResult <- triggerProposeF(casper, isAsync)
+            r <- proposerResult match {
+                  case ProposerEmpty =>
+                    logDebug(s"Failure: another propose is in progress")
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:112]: @zsluedem I fixed the message to stay in the same format. Can you point out where else we are depending on the returning text except in pyrchain?

Comment (zsluedem) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:112]: some of the integration tests depend on the logs. But the tests are passing. I think it is fine.

**Review by zsluedem — APPROVED** [review]
LGTM

**Comment by tgrospic** [comment]
bors merge

**Pull request merged**